File: java/src/main/java/io/github/securekit/RateLimiter.java
package io.github.securekit;

/**
 * Enforces a maximum number of actions per key within a rolling window.
 */
public final class RateLimiter {

    private final RateLimiterStore store;
    private final int limit;
    private final long windowMillis;

    public RateLimiter(RateLimiterStore store, int limit, long windowMillis) {
        this.store = store;
        this.limit = limit;
        this.windowMillis = windowMillis;
    }

    public boolean allow(String key) {
        return store.increment(key, windowMillis) <= limit;
    }
}

File: php/src/RateLimiter.php
<?php

declare(strict_types=1);

namespace SecureKit;

/**
 * Enforces a maximum number of actions per key within a rolling window.
 */
final class RateLimiter
{
    public function __construct(
        private readonly RateLimiterStoreInterface $store,
        private readonly int $limit,
        private readonly int $windowSeconds
    ) {
    }

    public function allow(string $key): bool
    {
        return $this->store->increment($key, $this->windowSeconds) <= $this->limit;
    }
}
